Slang

spy16 opened the pull request:
*Slang* (short for Sabre Lang)  - A tiny LISP dialect implemented using Sabre.

spy16 committed:
Add slang package with REPL, merge core into slang

spy16 committed:
Update README about Slang

spy16 commented:
@lthibault Any thoughts on this? I was thinking a package `slang` can have a simple usable language implementation with REPL, name-spaced scope using Sabre . This package can act as a reference usage.

spy16 committed:
Restructure slang internals

spy16 committed:
Better reflection for functions

lthibault commented:
Ah, wonderful!  This is a huge help in understanding how the pieces fit together 👍 

I'm going integrate some of the design of Slang into my own project and then post it to github.  If you're interested, it might provide some useful insight into how I'm using the library.

spy16 committed:
Add standard functions like gt, lt, gte etc.

spy16 committed:
Cleanup multiline prompt

spy16 committed:
Cleanup eval error handling in REPL

spy16 committed:
Implement Seq for string, fix / operator

spy16 committed:
Add functional options for REPL initialization

lthibault committed:
Add options

lthibault committed:
Print stack trace with log.Fatal call when setting up Slang instance

lthibault committed:
main application for Slang interpreter

lthibault committed:
Log stack trace when failing from slang.New()

lthibault committed:
Make REPL consume an interface instead of concrete *Slang.
Add WithBanner REPLOption

lthibault committed:
Remove Slang Option

lthibault committed:
Separate Slang runtime from REPL implementation.
This allows useres to write their own languages targeting a generic
sabre REPL implementation.  See cmd/slang/main.go for an example.

lthibault committed:
Merge slang

lthibault committed:
Factor readline dependency out of repl package.  Libreadline instance is
now passed to REPL through constructor args.

lthibault committed:
Remove github.com/lthibault/log

lthibault committed:
Update readme

lthibault committed:
Remove cmd/slang (reason: redundant with cmd/sabre)

lthibault committed:
don't print banner if empty

spy16 committed:
Merge pull request #7 from lthibault/lthibault

Separate Slang runtime from REPL implementation

spy16 committed:
Add error mapper to WithInput option

spy16 commented:
@lthibault take a look at https://github.com/spy16/sabre/pull/6/commits/679889252ed6020ac704dc76265dfb85c4f6763c. It implements our discussion. I went with `repl.Loop(ctx)` function. Any thoughts on that (i.e., `Run(ctx)` vs `Loop(ctx)`)?

lthibault commented on README.md:
```diff
@@ -17,12 +17,55 @@ Sabre is highly customizable, embeddable LISP engine for Go.
+  myRepl := repl.New(&myLang{})
+
+  for  {
+    if err := myRepl.Next(); err != nil {
```
We should update this example before merging.

spy16 replied:
I have already updated this. Pushed a few commits after this one.

spy16 committed:
Add basic print functions and fix nil form case in Eval

spy16 committed:
Update readme

lthibault commented on repl/option.go:
```diff
@@ -0,0 +1,84 @@
+
+// WithInput sets the REPL's input stream. `nil` defaults to bufio.Scanner
+// backed by os.Stdin
+func WithInput(in Input, errMapper func(error) error) Option {
```
Thoughts about making an `ErrorMapper` type?  I often find this easier to use because inline function signatures get noisy quite quickly. 

I'm inclined to do something like:

```go
type ErrorMap interface{
    Map(error) error
}

type ErrorMapFunc func(error) error

func (f ErrorMapFunc) Map(err error) error {
    return f(err)
}
```

This makes it easy for users to supply complex types for `errMap`.  A motivating example:

```go
logger := log.New()  // github.com/lthibault/log/pkg

errMap := func(err error) error {
    switch {
    case errors.Is(err, io.EOF):
         logger.Warn("exiting")
        return io.EOF
    case errors.Is(err, ww.Interrupt):
        logger.Trace("keyboard interrupt")
        return nil
    }    
}

r := repl.New(runtime,
    repl.WithInput(input, repl.ErrorMapFunc(errMap))
```

See [net/http::HandlerFunc](https://golang.org/pkg/net/http/#HandlerFunc) for a similar pattern in the Go standard library.

spy16 replied:
I agree. Will define a type for error-mapper.  But not sure if an interface type is needed or not. This mapper is expected to be a fixed mapping between an error to either `nil` or some other error. I can't think of any scenario where there would be some complex logic to decide this.

lthibault replied:
Ok fair enough.  So you're thinking `type ErrorMapper func(error) error` ?

Works for me 👍

spy16 replied:
Yep.

lthibault reviewed (dismissed):
LGTM overall.  Thoughts on the two comments I made?

spy16 commented:
While `repl` has turned out be a nice little package, i think the `Runtime` interface adds a bit of unnecessary complexity. For building a custom language, all that is really necessary with `Sabre` is to bind custom functions into a `Scope` implementation and pass this scope + forms to `sabre.Eval`. Historically also, LISP had concept of Environment against which data structures/forms/s-expressions are evaluated.  I am thinking may be it would be good enough to have `repl.New(env sabre.Scope, opts ...Option)` as the actual signature. Scope can implement an optional interface to support namespaces.

```golang
type NamespacedScope struct {
        SwitchNS(ns string)
        CurrentNS() string
}
```

This would make the REPL usage much simpler since user doesn't really have to manually implement any interfaces.. `repl.New(sabre.NewScope(nil), ....)` would work for most cases.

lthibault commented:
+1 for eschewing `repl.Runtime` in favor of `sabre.Scope`.  Also agree with the optional interface approach to providing namespaces.

spy16 committed:
Fix duplicate handling in 'set' function

spy16 committed:
Add WithReaderFactory REPL option for setting custom reader instance

spy16 commented:
@lthibault I have changed inline function to `ErrMapper` type and also added `WithReaderFactory` to allow customizing reader initialization. Let me know what you think..

Also, currently, binary is available as `sabre` (because of `cmd/sabre/main.go`). But given that the binary actually running `slang` as the runtime, do you think the binary should be available as `slang`  (i.e., `cmd/slang/main.go`) ?

lthibault commented:
>Do you think the binary should be available as `slang`?

Yes, I think that's a good idea.  I initially started a `cmd/slang` package because I hadn't noticed that `cmd/sabre` was doing what I wanted ... so I'm sure that will confuse others, too!

Reviewing changes now.

lthibault commented on repl/option.go:
```diff
@@ -0,0 +1,116 @@
+
+// ReaderFactory should return an instance of reader when called. This might
+// be called repeatedly. See WithReaderFactory()
+type ReaderFactory func(r io.Reader) *sabre.Reader
```
IMHO this should be an interface.  The logic to set up reader macros can easily become complex, especially when you're looking to re-use objects across Readers or when your language has primitives that deal with the network (as is my case with wetware).

For example, my `ReaderFactory` will undoubtedly be a struct that encapsulates a libp2p `host.Host` instance, which then needs to be bound to the `Anchor` macro.  It'll end up looking something like this:

```go
// Register the wetware Anchor macro.  This is directly copy-pasted
// from the existing ww source.
func Register(rd *sabre.Reader, root ww.Anchor) {
	rd.SetMacro('/', anchorMacro(root), false)
}

type factory struct {
    log log.Logger
    host host.Host  // libp2p host
}

func (f factory) NewReader(r io.Reader) *sabre.Reader {
    root := anchor.New(f.host, anchor.WithLogger(f.log))

    rd := sabre.NewReader(r)
    Register(rd, root)
    
    return rd
}
```
Obviously this _could_ be implemented without the `readerFactory` struct, but this is a case where I anticipate wanting the encapsulation that structs provide.

In any case, I'd really like to be able to write:

```go
repl.WithReaderFactory(factory{log: logger, host: h})
```

spy16 replied:
Ah, fair enough. I actually looked at `NewReader` and made the `ReaderFactory` to simply match that. I guess we could have `ReaderFactory` as interface and may be provide a `ReaderFactoryFunc` also for convinience.

lthibault replied:
Yup, that would be lovely.

lthibault commented on repl/repl.go:
```diff
@@ -0,0 +1,178 @@
+		return errors.New("scope is not set")
+	}
+
+	for {
```
Minor detail:  you can get rid of the select statement by writing:

```go
for ctx.Err() == nil {
    // read, eval, print ...
}
```

It removes one level of nesting, which is always nice.

spy16 replied:
Yes. I keep forgetting when there is `default` in `select` i don't really need to use `ctx.Done()`

lthibault commented on slang/core_test.go:
```diff
@@ -0,0 +1,91 @@
+package slang_test
```
Shouldn't tests live in the `slang` package?  My understanding is that the compiler doesn't include code from `*_test.go` files when `go build` or `go run` are invoked, only `go test`.

Keeping tests in the same package allows us to test unexported symbols.

spy16 replied:
I prefer to use `_test` package as much as possible for following reasons:
* Ideally all private code should reachable through the public interface of the package. Keeping tests in separate `_test` package allows you to ensure this.
* When using this pattern, you have access to only public interface of the package. So the tests are very similar to how a user of the package would use it. This allows to always keep in mind the public API of the package. Also, tests can now be used as reference also by users without really having to dig into internals.

lthibault replied:
Fine by me - those are sensible reasons (especially the second point) 😃

lthibault reviewed (commented):
LGTM overall.  I've left a few comments.

spy16 committed:
Modify ReaderFactory to be interface, simplify repl.Loop

spy16 commented:
All fixed. I think it is good time to merge this to master and continue from there.

lthibault commented:
@spy16 can we increment the version tag too?

spy16 commented:
Yes will do that. `v0.2.0` ? (Still not mature enough to be `v1` i think. But big changes so minor release)

lthibault commented:
> Yes will do that. `v0.2.0` ? (Still not mature enough to be `v1` i think. But big changes so minor release)

Yes, agreed on all counts.

spy16 committed:
Move sabre binary to slang

lthibault commented:
BTW, I think this is ready to merge ... just want to make sure you weren't waiting on me! 😄

spy16 commented:
Ah thanks. But it was late night for me and decided to do it later. Will do it sometime today.

spy16 commented:
I want to cleanup the commits here a little bit before merging since there are lot of unwanted/experimental commits. Any ideas? Interactive re-base is getting lot of conflicts for some reason.

One other option is to squash and merge, but that will lose author information unless we add `Co-authored-by`

lthibault commented:
> I want to cleanup the commits here a little bit before merging since there are lot of unwanted/experimental commits. Any ideas? Interactive re-base is getting lot of conflicts for some reason.
> 
> One other option is to squash and merge, but that will lose author information unless we add `Co-authored-by`

Sorry, I was a bit slow to respond.

I'm generally a fan of rebasing into the master branch to the extent possible.  If this situation ever comes up again, let's just discuss again at that point (and hope it doesn't happen to often!)

In any case, I very much appreciate your consideration of attribution.  Thanks for that 😃 

(P.S.:  first collaborative PR closed! Woohoo! 🎉)

spy16 commented:
Yea i like to rebase onto master as well. But before that i also usually cleanup commits a bit by squashing some of them to make each commit a sensible and usable set of changes.

For this PR, i have included the `Co-authored-by` section which shows up in this merge commit. But unfortunately it doesn't show in repository contributor list.. :(

lthibault commented:
No worries.  There will be a next time 😃
